src/metabase/driver/mysql.clj

```diff
@@ -348,6 +353,40 @@
                                 2)
                           (hx/literal "-01"))))
 
+(defmethod sql.qp/->honeysql [:mysql :datetime-diff]
+  [driver [_ x y unit :as clause]]
+  (let [x (sql.qp/->honeysql driver x)
+        y (sql.qp/->honeysql driver y)
+        disallowed-types (keep
+                          (fn [v]
+                            (when-let [db-type (some-> v hx/type-info hx/type-info->db-type keyword str/upper-case)]
+                              (let [base-type (sql-jdbc.sync/database-type->base-type driver db-type)]
+                                (when-not (some #(isa? base-type %) [:type/Date :type/DateTime])
+                                  (name db-type)))))
+                          [x y])]
+    (when (seq disallowed-types)
+      (throw (ex-info (tru "Only datetime, timestamp, or date types allowed. Found {0}"
+                           (pr-str disallowed-types))
+                      {:found disallowed-types
+                       :type  qp.error-type/invalid-parameter})))
```

Comment:
I was thinking in would be nice if this could be abstracted out somewhere so we could reuse it between the different drivers, or at least if some of the stuff implementation stuff could be abstracted out 

For example the type-info -> db-type -> base-type stuff could be a good thing to put in some general util namespace for `:sql-jdbc` drivers. But we actually don't have any appropriate place to put it currently I think. So we'd have to make a new namespace.

I guess it's ok for now unless you want to make some sort of `:sql-jdbc` util namespace to put this in

Reply:
Yes, it would be nice to have a kind of static type checker for MBQL. The MBQL schema is already doing some type checking because it knows the argument and return types of some expressions, but it doesn't know the type of fields. Perhaps there's a way to integrate field types into the schema definition.

Regardless, once I've implemented datetime-diff for a few more drivers I'll strip out this repeated stuff into a util namespace like you suggested.